src/Type/Php/RegexShapeMatcher.php

```diff
@@ -0,0 +1,181 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use Hoa\Compiler\Exception\Exception;
+use Hoa\Compiler\Llk\Llk;
+use Hoa\Compiler\Llk\Parser;
+use Hoa\Compiler\Llk\TreeNode;
+use Hoa\File\Read;
+use PHPStan\Analyser\TypeSpecifierContext;
+use PHPStan\Type\Constant\ConstantArrayTypeBuilder;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\IntegerRangeType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function array_key_last;
+use function array_keys;
+use function in_array;
+use function is_int;
+use function is_string;
+use function preg_match;
+use function preg_replace;
+use function str_contains;
+use const PREG_OFFSET_CAPTURE;
+use const PREG_UNMATCHED_AS_NULL;
+
+final class RegexShapeMatcher
+{
+
+	private static ?Parser $parser = null;
+
+	/**
+	 * @param int-mask<PREG_OFFSET_CAPTURE|PREG_UNMATCHED_AS_NULL>|null $flags
+	 */
+	public function matchType(string $regex, ?int $flags, TypeSpecifierContext $context): ?Type
+	{
+		// add one capturing group to the end so all capture group keys
+		// are present in the $matches
+		// see https://3v4l.org/sOXbn, https://3v4l.org/3SdDM
+		$captureGroupsRegex = preg_replace('~.[a-z\s]*$~i', '|(?<phpstanNamedCaptureGroupLast>)$0', $regex);
+
+		if (
+			$captureGroupsRegex === null
+			|| @preg_match($captureGroupsRegex, '', $matches, PREG_UNMATCHED_AS_NULL) === false
+		) {
+			return null;
+		}
+		unset($matches[array_key_last($matches)]);
+		unset($matches['phpstanNamedCaptureGroupLast']);
+
+		$remainingNonOptionalGroupCount = $this->countNonOptionalGroups($regex);
+		if ($remainingNonOptionalGroupCount === null) {
+			// regex could not be parsed by Hoa/Regex
```

Comment:
I would prefer emitting a PHPStan error with the parse exception message - the PCRE grammar is actually not so complex, so we can fix the Hoa grammar based on issue reports.

This would effectively also validate constant regexps for syntax errors.

The https://3v4l.org/sOXbn parsing hack can be then removed in favor of Hoa parser.